mds/provider_poller/poller.py

```diff
@@ -196,8 +215,120 @@ def _poll_status_changes(self):
+                getattr(settings, "PROVIDER_POLLER_LIMIT", 90)
+            )
+
+        # But we now apply a "lag" before actually polling,
+        # leaving time for the provider to collect data from its devices
+        polling_lag = self.provider.api_configuration.get("provider_polling_lag")
+        if polling_lag:
+            polling_lag = parse_duration(polling_lag)
+            if (timezone.now() - last_event_time_polled) < polling_lag:
+                logger.info("Still under the polling lag, back to sleep.")
+                return
+
+        # The MDS 0.4 Provider API got two endpoints:
+        # - /events for the real-time or so data (formerly /status_changes)
+        #   but limited to two weeks of history
+        # - /status_changes for all the data except the current hour
+        # If we're catching up far in time, begin by polling /status_changes
+        realtime_threshold = parse_duration(  # Default is 9 days
+            self.provider.api_configuration.get("realtime_threshold", "P9D")
+        )
+        if (timezone.now() - last_event_time_polled) > realtime_threshold:
+            # We have to query the archived status changes with another format
+            logger.info("last_event_time_polled is too old, asking archives")
+            params["event_time"] = last_event_time_polled.isoformat()[
+                : len("YYYY-MM-DDTHH")
+            ]
+        else:
+            # Query the real-time endpoint as usual
+            params["start_time"] = utils.to_mds_timestamp(
+                self.from_cursor or last_event_time_polled
+            )
+
+        # Provider-specific params to optimise polling
+        try:
+            params.update(self.provider.api_configuration["status_changes_params"])
+        except KeyError:
+            pass
+
+        endpoint = "events"  # The new name for the real-time events endpoint
+        if "event_time" in params:
+            # We asked the archived status changes instead
+            endpoint = "status_changes"
+        next_url = urllib.parse.urljoin(self.provider.base_api_url, endpoint)
+        if self.provider.api_configuration.get("trailing_slash"):
+            next_url += "/"
+
+        if params:
+            next_url = "%s?%s" % (next_url, urllib.parse.urlencode(params))
+
+        # Pagination
+        while next_url:
+            body = self._get_body(next_url, api_version)
+            # Translate older versions of data
+            translated_data = translate_data(body["data"], body["version"])
+            status_changes = translated_data["status_changes"]
+            next_url = body.get("links", {}).get("next")
+
+            # A transaction for each "page" of data
+            with transaction.atomic():
+                if status_changes:
+                    # We get the maximum values from the status changes
+                    event_time_polled, _ = self._process_status_changes(status_changes)
```

Comment:
Fetching is version dependant, but the internal processing is only compatible with the latest version, hence the translation of the body beforehand.